app/src/test/java/io/appium/uiautomator2/utils/w3c/W3CActionsPreprocessingTests.java

```diff
@@ -175,7 +176,29 @@ public void verifyValidActionChainPreprocessingWithCancelChange() throws JSONExc
                 "{\"type\": \"pointerMove\", \"duration\": 1000, \"origin\": \"pointer\", \"x\": -50, \"y\": 0}," +
                 "{\"type\": \"pointerUp\", \"button\": 0}]" +
                 "} ]");
-        assertThat(preprocessActions(actionJson).toString(), is(equalTo(processedJson.toString())));
+        assertThat(actionsPreprocessor.preprocess(actionJson).toString(), is(equalTo(processedJson.toString())));
+    }
+
+    @Test(expected = ActionsParseException.class)
+    public void verifyChainPreprocessingFailsIfMultiplePointerTypesArePresent() throws JSONException {
+        final JSONArray actionJson = new JSONArray("[ {" +
+                "\"type\": \"pointer\"," +
```

Comment:
Ugh. Does Java have nothing like a here-doc?

Reply:
unfortunately not. They've just added something like that into java10 spec